contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -297,46 +315,53 @@ contract FourYearVestingLogic is IFourYearVesting, FourYearVestingStorage, Appro
      * to an address specified by the token owner. Low level function.
      * @dev Once here the caller permission is taken for granted.
      * @param receiver The receiving address.
-     * @param isGovernance Whether all tokens (true)
      * or just unlocked tokens (false).
      * */
-    function _withdrawTokens(address receiver, bool isGovernance) internal {
+    function _withdrawTokens(address receiver, uint256 startFrom) internal {
         require(receiver != address(0), "receiver address invalid");
 
         uint96 stake;
 
         /// @dev Usually we just need to iterate over the possible dates until now.
         uint256 end;
 
+        uint256 defaultStart = startDate.add(extendDurationFor);
+
+        startFrom = startFrom >= defaultStart ? startFrom : defaultStart;
+
         /// @dev In the unlikely case that all tokens have been unlocked early,
         ///   allow to withdraw all of them.
-        if (staking.allUnlocked() || isGovernance) {
+        if (staking.allUnlocked()) {
             end = endDate;
         } else {
             end = block.timestamp;
         }
 
+        uint256 totalIterationValue =
+            (startFrom + (FOUR_WEEKS * getMaxVestingWithdrawIterations()));
+        uint256 adjustedEnd = end < totalIterationValue ? end : totalIterationValue;
+
         /// @dev Withdraw for each unlocked position.
         /// @dev Don't change FOUR_WEEKS to TWO_WEEKS, a lot of vestings already deployed with FOUR_WEEKS
         ///		workaround found, but it doesn't work with TWO_WEEKS
         /// @dev For four year vesting, withdrawal of stakes for the first year is not allowed. These
         /// stakes are extended for three years. In some cases the withdrawal may be allowed at a different
         /// time and hence we use extendDurationFor.
-        for (uint256 i = startDate.add(extendDurationFor); i <= end; i += FOUR_WEEKS) {
+        for (uint256 i = startFrom; i < adjustedEnd; i += FOUR_WEEKS) {
```

Comment:
```suggestion
for (uint256 i = startFrom; i <= adjustedEnd; i += FOUR_WEEKS) {
```  
the same reason as for VestingLogic and keeping code duplication already creates hassle  
please consider a quick refactoring - extract duplicate code into a separate abstract contract

Reply:
updated